Start adding support for Mongodb Atlas database

alkampfergit opened the pull request:
This is a first implementation where basic unit test passes and it was tested with
some custom code. It works with cluster from M10 upward (does not work still with shared cluster).

Still have some perplexity on how to use minRelevance because I did not checked if this is a -1 1 value (cosine distance) or whatever.

This code contains also the necessary code to automatically create the Vector Indexes needed to execute queries.

Please let me know if I need to fulfill special check for the PR to be closed. Also if you think that a concrete implementation should be placed into my repository and does not find space in the official one you can tell me and reject the PR, I'll move the code in a repository of mine.

<!-- Thank you for your contribution! Please provide the following information -->
I wanted support for mongodb Atlas, I've already done a pull request but it was based on dev branch that gots deleted, so I reopened on main branch.


## High level description (Approach, Design)

dluc commented:
hi @alkampfergit thanks for the PR! 

Couple of asks:

1. could you add a script under /tools to start Mongodb Atlas with docker (assuming that's possible). See other scripts in the folder
2. about correctness, I would suggest extending these tests to compare results:
  * https://github.com/microsoft/kernel-memory/blob/main/service/tests/Core.FunctionalTests/VectorDbComparison/TestCosineSimilarity.cs. (you can use the booleans to compare just Mongodb vs SimpleVectorDb)
  * https://github.com/microsoft/kernel-memory/blob/main/service/tests/Core.FunctionalTests/VectorDbComparison/TestMemoryFilters.cs

dluc commented on extensions/AtlasMongoDb/AtlasMongoDb.FunctionalTests/appsettings.json:
```diff
@@ -0,0 +1,57 @@
+  },
+  "Services": {
+    "MongoDb": {
+      "ConnectionString": "",
```
could you add a comment with an example connection string?

alkampfergit commented:
> hi @alkampfergit thanks for the PR!
> 
> Couple of asks:
> 
> 1. could you add a script under /tools to start Mongodb Atlas with docker (assuming that's possible). See other scripts in the folder
> 2. about correctness, I would suggest extending these tests to compare results:
> 
> * https://github.com/microsoft/kernel-memory/blob/main/service/tests/Core.FunctionalTests/VectorDbComparison/TestCosineSimilarity.cs. (you can use the booleans to compare just Mongodb vs SimpleVectorDb)
> * https://github.com/microsoft/kernel-memory/blob/main/service/tests/Core.FunctionalTests/VectorDbComparison/TestMemoryFilters.cs

Thanks a lot for the instructions, I'll proceed to implement these tests to verify that everything is ok. For what regard docker I'll investigate: Atlas is a cloud only product, but it has a commandline that can start a local cluster for tests, but I really didn't tried vector capabilities. 

I'll proceed to check and I'll update the PR accordingly.

Thanks!

alkampfergit commented:
Hi @dluc 

I've implement those two tests you suggeested in the previous comment. Sadly enough MongoDb perform a normalization of the base Cosine Similarity so I needed to recalcolate in memory the correct value for the test to pass. 

I've added a Markdown file explaining how to run a local Atlas MongoDB installation locally thanks to Atlas CLI. It uses podman under the hood, but it cannot work with docker. This is th easiest way to have a local running instance of AtlasDB to perform tests.

I'm waiting for comments in the PR if Any.

Thanks.

dluc commented on extensions/AtlasMongoDb/AtlasMongoDb/MongoDbVectorMemory.cs:
```diff
@@ -0,0 +1,472 @@
+        string index,
+        string text,
+        ICollection<MemoryFilter>? filters = null,
+        double minRelevance = 0,
```
looks like this is never used, does the method filter out records by relevance if a user passes a value greater than 0?

alkampfergit replied:
We have a problem here, I can try to implement this but atlas does not seems to filter for cosine similarity value, because it internally normalize the score ...  The only quick way that I can do is filtering client side, stopping the enumeration when the value falls below the relevance.

Also many vectror db like pinecone and also atlas mongodb can compute similiaryt with standard euclidean distance or cosine similarity or dot product (in case of unary vectors) so I'm not sure how to interpret that minRelevance score. If you ensure me that it is always pure cosine similiarity I can easily filter out during enumeration. Thanks

dluc replied:
Thanks, I will take a look soon. About similarity, currently the expectation is to always use Cosine Similarity and to have the same value across all the engines. There was some thought to include other metrics in future, via configuration, but the default should be the same, so users can switch storage without having having to think about the math :-)

alkampfergit replied:
I completely agree, sadly mongodb does some normalization, but in the end you can still filter in memory. 😃

alkampfergit replied:
Also I recalculate pure cosine similarity in memory after query, from this point of view the client will have consistent numbers.

dluc commented on tools/run-atlas.sh:
nit: please rename to `run-mongodb-atlas.sh` and set the perms to 0750 (executable)

alkampfergit replied:
done

dluc commented on service/tests/Core.FunctionalTests/DefaultTestCases/DocumentUploadTest.cs:
```diff
@@ -16,10 +16,18 @@ public static async Task ItUploadsPDFDocsAndDeletes(IKernelMemory memory, Action
+
+            //Avoid stalling forever, whathever bug could be in the database/pipeline this test must exit
+            //after a reasonable amount of time
+            if (DateTime.UtcNow.Subtract(startWait).TotalSeconds > 120)
```
nit: pls use `DateTimeOffset` class

dluc commented on extensions/AtlasMongoDb/AtlasMongoDb/Helpers/AtlasSearchHelper.cs:
```diff
@@ -0,0 +1,362 @@
+/// </ul>
+/// </para>
+/// </summary>
+public class AtlasSearchHelper
```
can we keep this class `internal` only?

alkampfergit replied:
I added internalsvisible to because it is used by test to cleanup.

dluc commented on extensions/AtlasMongoDb/AtlasMongoDb/MongoDbKernelMemoryStorage.cs:
```diff
@@ -0,0 +1,216 @@
+        return Task.CompletedTask;
+    }
+
+    private static async Task<IAsyncCursor<GridFSFileInfo<string>>> GetFromBucketByIdAsync(string id, GridFSBucket<string> bucket, CancellationToken cancellationToken)
```
here and in other files, could you sort methods by visibility?
This is the order we try to keep across files (with few exceptions):
* public constants values / static values (possibly no public constants, they are all grouped in one file)
* private constants / static values
* public getters/setters
* ctors
* public methods
* public static methods
* private methods
* private static methods

alkampfergit replied:
I should have fixed ordering.

dluc commented on extensions/AtlasMongoDb/AtlasMongoDb/ConfigurationHelper.cs:
nit: rename file to `DependencyInjection.cs`

alkampfergit replied:
done

dluc commented on extensions/AtlasMongoDb/AtlasMongoDb/ConfigurationHelper.cs:
```diff
@@ -0,0 +1,59 @@
+    /// </summary>
+    /// <param name="builder">The kernel builder</param>
+    /// <param name="configuration">Configuration for Mongodb</param>
+    public static IKernelMemoryBuilder WithAtlasMemoryDb(
```
as a user I might want to use MongoDb Atlas only for memory records, keeping files on disk. Could you add more extension methods to allow that?

dluc replied:
I would also rename the method, similar to other places. E.g. here "Atlas" I think should be "MongoDbAtlas"

alkampfergit replied:
Done

SimonLuckenuik commented:
Thanks for the work @alkampfergit , any chance this could be adapted to work with Azure MongoDB for vCore?
Maybe something like a base class MongoDb, and one specialized for Atlas and one for MongodDB vCore with their specific needs?
I believe the index creation and search syntax would be different, but nothing more:
https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search#use-the-createindexes-template-to-create-a-vector-index

alkampfergit commented on extensions/AtlasMongoDb/AtlasMongoDb/MongoDbKernelMemoryStorage.cs:
```diff
@@ -0,0 +1,216 @@
+
+namespace Alkampfer.KernelMemory.AtlasMongoDb;
+
+public class MongoDbKernelMemoryStorage : MongoDbKernelMemoryBaseStorage, IContentStorage
```
Actually MongoDb has GridFs to store files and blob. In production system I do not like to have all the blob and information on disk, also I can benefit from the ability to query from various UI interface all the record inside my mongodb for diagnostic. Clearly if you want to use only the MemoryDb part you can simply register only the MEmoryDb part completely ignoring this class.

alkampfergit commented:
> Thanks for the work @alkampfergit , any chance this could be adapted to work with Azure MongoDB for vCore? Maybe something like a base class MongoDb, and one specialized for Atlas and one for MongodDB vCore with their specific needs? I believe the index creation and search syntax would be different, but nothing more: https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search#use-the-createindexes-template-to-create-a-vector-index

Actually we are using MongoDbAtlas, I do not know MongodDB vCore, does it support Vector Search in the same way atlas does? (lucene index).

alkampfergit commented on extensions/AtlasMongoDb/AtlasMongoDb/AtlasMongoDb.csproj:
```diff
@@ -0,0 +1,20 @@
+        <TargetFramework>net6.0</TargetFramework>
+        <RollForward>LatestMajor</RollForward>
+        <AssemblyName>Alkampfer.KernelMemory.AtlasMongoDb</AssemblyName>
+        <RootNamespace>Alkampfer.KernelMemory.AtlasMongoDb</RootNamespace>
```
Done

alkampfergit commented on tools/create-local-atlas.md:
done

alkampfergit commented:
@dluc I was pretty busy these days, but I should had resolved all the issues. Feel free to request additional modification (i've left two commits, the first is the original for the PR, the other contains all modification done as for your request)

alkampfergit commented on extensions/AtlasMongoDb/AtlasMongoDb.FunctionalTests/DefaultTests.cs:
```diff
@@ -0,0 +1,111 @@
+using Microsoft.KernelMemory;
+using Microsoft.TestHelpers;
+
+namespace AtlasMongoDb.FunctionalTests;
```
Done

SimonLuckenuik commented:
> > Thanks for the work @alkampfergit , any chance this could be adapted to work with Azure MongoDB for vCore? Maybe something like a base class MongoDb, and one specialized for Atlas and one for MongodDB vCore with their specific needs? I believe the index creation and search syntax would be different, but nothing more: https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search#use-the-createindexes-template-to-create-a-vector-index
> 
> Actually we are using MongoDbAtlas, I do not know MongodDB vCore, does it support Vector Search in the same way atlas does? (lucene index).

Yes it does support vector search, but the index and query syntaxes are a bit different, see here: https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search#use-the-createindexes-template-to-create-a-vector-index

Someone also started a MongoDB vCore pull request: https://github.com/microsoft/kernel-memory/pull/329

In the end, for maintenance, most probably it would be more interesting to have a common shared code base.

dviry commented:
Sorry for meddling, but I have been using this and I have some feedback:
- why the requirement to use Mongo v7? Most Atlas clusters (except the highest plans) cannot yet be upgraded to v7
- the [knnBeta search type is deprecated](https://www.mongodb.com/docs/atlas/atlas-search/knn-beta/), it should use the normal $vectorSearch
- there are some warnings from the NetAnalyzers regarding code-quality (#nullable)
- why hard-coding BsonDocuments everywhere instead of using the standard .NET driver? Would make it more future-proof
- the collectionName is hard coded (even the ConnectionNamePrefix cannot be changed from the default "_ix_") I would like to be able to configure that (we are using a naming convention within one Mongo db)
- I haven't seen any error-handling - should all exceptions always bubble up? In my experience the MongoDB exceptions are not always helpful and the logging might have issues with those
- it's not really a best-practice to store long strings in the Mongo _id field - this has a performance/storage impact. I also see string storage in tg___file_id[] - although it's a GUID that could be well stored in a MongoDB iuid/binary field

Just a general question - I know SemanticKernel is a different project but the connectors are very similar - couldn't we somehow share/reuse some of the codebase (or just have a wrapper?), for example [the MongoDB Connector](https://github.com/microsoft/semantic-kernel/tree/main/dotnet/src/Connectors/Connectors.Memory.MongoDB)

alkampfergit commented:
UPDATE: I've updated the pull request as for @dviry request, for now I've addressed support for 6 and 7 (tested with local atlas insetallation), also removed knnBeta to favor $vectorSearch.

Still the code can be improved to let the user choose naming of the collection and other minor updates.

Fixed also a problem in creation of Vector Search index where a stale vector resulted as created but was not functional.

dviry commented:
Thanks for considering my feedback!

One question before I update my local copy - am I missing something or is the [knnBeta](https://github.com/microsoft/kernel-memory/blob/ccf8fcfe62e509e5d85bf42cdd6322baa8fb2e23/extensions/MongoDbAtlas/MongoDbAtlas/MongoDbVectorMemory.cs#L369) still in the latest code?

Thanks

dviry commented on extensions/MongoDbAtlas/MongoDbAtlas/Helpers/MongoDbAtlasSearchHelper.cs:
```diff
@@ -0,0 +1,303 @@
+        {
+            new() {
+                {
+                    "$listSearchIndexes",
```
note that this only works starting M10 instances (not on "free forever" M0)

alkampfergit replied:
I know, do you know a way to get the list of indexes in an M0 instance? The only workaround I know is an option where you avoid the code checking and creating index and you need to create index manually.

Alternatively, I can simply try to create the index and trap exception of index already existing so it can work even in M0.

dviry replied:
AFAIK that won't work either, since creating SearchIndexes is also only available starting M10...

alkampfergit replied:
The only solution would be adding an option to the configuration telling the component not to create the indexes, so when the core code of Kernel Memory tells you to create an index, you will simply skip.

dviry replied:
I personally prefer less hidden "options" - I would rather handle that specific error (both on index check/creation) and log it accordingly. Additionally I would add it as a requirement to the connector Documentation (maybe with details on how to create it manually in case anyone wants to keep using an M0 instance for development).

alkampfergit replied:
I'll check, I had a situation where the collection was created but the index no, this is the reason why I always check for the index. The UpdateOneAsync can be avoided, even if I thought that CreateIndexAsync() is a method that is rarely called, I'll check as soon as possible, thanks.

dviry commented on extensions/MongoDbAtlas/MongoDbAtlas/Helpers/MongoDbAtlasSearchHelper.cs:
```diff
@@ -0,0 +1,303 @@
+                    new BsonDocument
+                    {
+                        { "name", this.GetIndexName(collectionName) },
+                        { "type", "vectorSearch" },
```
type should be ```vector```

alkampfergit commented:
> Thanks for considering my feedback!
> 
> One question before I update my local copy - am I missing something or is the [knnBeta](https://github.com/microsoft/kernel-memory/blob/ccf8fcfe62e509e5d85bf42cdd6322baa8fb2e23/extensions/MongoDbAtlas/MongoDbAtlas/MongoDbVectorMemory.cs#L369) still in the latest code?
> 
> Thanks
> 
> EDIT: sorry my mistake - the method CreatePipeline() is still there but not referenced anywhere

Removed and amended last commit, I've left old code for review and compare but it makes sense to remove to have cleaner code.

dluc commented:
For some reason the service doesn't work on my workstation, e.g. the cosine similarity test is unable to connect.

```
⚡️ ./run-mongodb-atlas.sh

[Default Settings]
Deployment Name   local6786
MongoDB Version   7.0
Port              27017

Creating your cluster local6786 [this might take several minutes]
1/3: Starting your local environment...
2/3: Downloading the MongoDB binaries to your local environment...
3/3: Creating your deployment local6786...
Deployment created!
Connection string: mongodb://root:root@localhost:27017/?directConnection=true

connection skipped
```

```
Deleting index test-cosinesimil in Microsoft.KernelMemory.MongoDbAtlas.MongoDbVectorMemory

System.TimeoutException
A timeout occurred after 10000ms selecting a server using CompositeServerSelector{ Selectors = ReadPreferenceServerSelector{ ReadPreference = { Mode : Primary } }, LatencyLimitingServerSelector{ AllowedLatencyRange = 00:00:00.0150000 }, OperationsCountServerSelector }. Client view of cluster state is { ClusterId : "1", DirectConnection : "True", Type : "Standalone", State : "Disconnected", Servers : [{ ServerId: "{ ClusterId : 1, EndPoint : "Unspecified/localhost:27017" }", EndPoint: "Unspecified/localhost:27017", ReasonChanged: "Heartbeat", State: "Disconnected", ServerVersion: , TopologyVersion: , Type: "Unknown", HeartbeatException: "MongoDB.Driver.MongoConnectionException: An exception occurred while opening a connection to the server.
 ---> System.Net.Sockets.SocketException (61): Connection refused [::1]:27017
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Connect(EndPoint remoteEP)
   at MongoDB.Driver.Core.Connections.TcpStreamFactory.Connect(Socket socket, EndPoint endPoint, CancellationToken cancellationToken)
   at MongoDB.Driver.Core.Connections.TcpStreamFactory.CreateStream(EndPoint endPoint, CancellationToken cancellationToken)
   at MongoDB.Driver.Core.Connections.BinaryConnection.OpenHelper(CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
```

> "ConnectionString": "mongodb://localhost:27017/?directConnection=true&serverSelectionTimeoutMS=10000",

dluc commented:
I would rename these classes, to be consistent with others:

* `MongoDbVectorMemory` => `MongoDbAtlasMemoryDb`
* `MongoDbKernelMemoryStorage` => `MongoDbAtlasStorage` or `MongoDbAtlasContentStorage`

![image](https://github.com/microsoft/kernel-memory/assets/371009/1fa72fd5-d180-4d47-801e-52588a878cfc)

dviry commented:
As mentioned above those commands only work starting M10 Atlas instances,
not on M0 ones... You will have to create the index manually in case you
want to continue using the M0 instance.

On Fri, Mar 8, 2024 at 2:03 PM Devis Lucato ***@***.***>
wrote:

> For some reason the service doesn't work on my workstation, e.g. the
> cosine similarity test is unable to connect.
>
> ⚡️ ./run-mongodb-atlas.sh
>
> [Default Settings]
> Deployment Name   local6786
> MongoDB Version   7.0
> Port              27017
>
> Creating your cluster local6786 [this might take several minutes]
> 1/3: Starting your local environment...
> 2/3: Downloading the MongoDB binaries to your local environment...
> 3/3: Creating your deployment local6786...
> Deployment created!
> Connection string: ***@***.***:27017/?directConnection=true
>
> connection skipped
>
> Deleting index test-cosinesimil in Microsoft.KernelMemory.MongoDbAtlas.MongoDbVectorMemory
>
> System.TimeoutException
> A timeout occurred after 10000ms selecting a server using CompositeServerSelector{ Selectors = ReadPreferenceServerSelector{ ReadPreference = { Mode : Primary } }, LatencyLimitingServerSelector{ AllowedLatencyRange = 00:00:00.0150000 }, OperationsCountServerSelector }. Client view of cluster state is { ClusterId : "1", DirectConnection : "True", Type : "Standalone", State : "Disconnected", Servers : [{ ServerId: "{ ClusterId : 1, EndPoint : "Unspecified/localhost:27017" }", EndPoint: "Unspecified/localhost:27017", ReasonChanged: "Heartbeat", State: "Disconnected", ServerVersion: , TopologyVersion: , Type: "Unknown", HeartbeatException: "MongoDB.Driver.MongoConnectionException: An exception occurred while opening a connection to the server.
>  ---> System.Net.Sockets.SocketException (61): Connection refused [::1]:27017
>    at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
>    at System.Net.Sockets.Socket.Connect(EndPoint remoteEP)
>    at MongoDB.Driver.Core.Connections.TcpStreamFactory.Connect(Socket socket, EndPoint endPoint, CancellationToken cancellationToken)
>    at MongoDB.Driver.Core.Connections.TcpStreamFactory.CreateStream(EndPoint endPoint, CancellationToken cancellationToken)
>    at MongoDB.Driver.Core.Connections.BinaryConnection.OpenHelper(CancellationToken cancellationToken)
>    --- End of inner exception stack trace ---
>
> "ConnectionString":
> "mongodb://localhost:27017/?directConnection=true&serverSelectionTimeoutMS=10000",
>
> I created an account and deployed a free instance at mongodb.net
>
> "ConnectionString": "mongodb+srv://<user>:<password>@
> kernelmemory.cfuqvtp.mongodb.net/?retryWrites=true&w=majority&appName=kernelmemory
> ",
>
> When running the cosine similarity functional test I get this error:
>
> MongoDB.Driver.MongoCommandException
> Command aggregate failed: $listSearchIndexes is not allowed or the syntax is incorrect, see the Atlas documentation for more information.
>    at MongoDB.Driver.Core.WireProtocol.CommandUsingCommandMessageWireProtocol`1.ProcessResponse(ConnectionId connectionId, CommandMessage responseMessage)
>
> Any suggestion?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/kernel-memory/pull/311#issuecomment-1985658833>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJMLNAJHM2FMDQOCMJMG6LYXGZIRAVCNFSM6AAAAABDGBKWHCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBVGY2TQOBTGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dluc commented:
> As mentioned above those commands only work starting M10 Atlas instances, not on M0 ones... You will have to create the index manually in case you want to continue using the M0 instance.

I see, so we need to figure out how to fix the local docker instance, otherwise we won't be able to test the integration. Unless we get support from MongoDb and they could offer a M10 instance for test purposes.

alkampfergit commented:
> > As mentioned above those commands only work starting M10 Atlas instances, not on M0 ones... You will have to create the index manually in case you want to continue using the M0 instance.
> 
> I see, so we need to figure out how to fix the local docker instance, otherwise we won't be able to test the integration. Unless we get support from MongoDb and they could offer a M10 instance for test purposes. In the meantime I'll try the version available via homebrew, see if it can support the integration tests.

From what I see you are connecting in port 27017, that is the standard mongodb port, but the script to start the cluster starts in docker and expose the port in 27777 so the connection string is

mongodb://root:root@localhost:27777/?authSource=admin

I've added a comment in the file. The reason behind this is that usually, developers that works with mognodb, have a standard local instance, then when they want to test atlas, they use a different port to avoid conflict.

To simplify everything we can change the port in the sh file to use standard one.

dluc commented:
Update: local tests via homebrew are working.
I made some changes to functional tests and renamed the "MongoDb" config key to "MongoDbAtlas".
All functional tests seem to pass fine.

alkampfergit commented:
> Update: local tests via homebrew are working. I made some changes to functional tests and renamed the "MongoDb" config key to "MongoDbAtlas". All functional tests seem to pass fine.

Good, sorry for the inconvenience, I was using secret because I tested against a M20 of a customer ,so I really do not want to leak the connectino string and I forgot to change the default one to match that one of the script .sh.

dluc commented on extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasConfig.cs:
```diff
@@ -0,0 +1,70 @@
+        return this;
+    }
+
+    internal IMongoDatabase GetDatabase()
```
this method seems a bit odd, e.g. having a client inside the config class. The code and the client instance could be moved to the `MongoDbAtlasBaseStorage` class.

alkampfergit replied:
it could be done, but it is best practice to have a single Client for the entire application, so it is standard practice to leave the client in a single instance of a global class. Actually it would be also better (maybe in a subsequent change) to give the caller the ability to specify the client class from external code.


I've moved inside a dedicated class.

dluc replied:
the memory db object is a singleton, that should be sufficient no?

dluc commented on extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasMemory.cs:
```diff
@@ -0,0 +1,351 @@
+    {
+        if (limit <= 0)
+        {
+            limit = 10;
```
this var is never used, the method should stop after returning `limit` records

alkampfergit replied:
I've put a check inside the loop, but this value is passed to the method that created the filter, the number of item is directly inserted in the mongodb query.

dluc commented on extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasMemory.cs:
```diff
@@ -0,0 +1,351 @@
+        // need to recompute cosine similarity using the embeddings.
+#if DEBUG
+        //List<(double, double, double)> similarities = new();
+#endif
```
could you remove the code commented out?

Looks like there are a bunch of #DEBUG blocks that could be removed across the file

dluc commented on extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasMemory.cs:
```diff
@@ -0,0 +1,351 @@
+        foreach (var filter in filters ?? Array.Empty<MemoryFilter>())
+        {
+            var thisFilter = filter.GetFilters().ToArray();
+            var numOfFilters = thisFilter.Count(x => !string.IsNullOrEmpty(x.Value));
```
this var is never used

alkampfergit replied:
you are right, it is a leftover for some tests I made at the beginning.

dluc commented on extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasMemory.cs:
```diff
@@ -0,0 +1,351 @@
+                filtersArray.Add(condition);
+            }
+
+            // All these filter if more than one must be enclosed in an end filter
```
could you revisit the comment, might be missing some word or have a typo - thanks

dluc commented on extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasStorage.cs:
```diff
@@ -0,0 +1,216 @@
+    public async Task DeleteIndexDirectoryAsync(string index, CancellationToken cancellationToken = new CancellationToken())
+    {
+        // Delete all document in GridFS that have index as metadata
+        var filter = Builders<GridFSFileInfo>.Filter.Eq("metadata.index", index);
```
var never used

alkampfergit commented:
Yes

Il dom 10 mar 2024, 19:26 Devis Lucato ***@***.***> ha
scritto:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasConfig.cs
> <https://github.com/microsoft/kernel-memory/pull/311#discussion_r1518925331>
> :
>
> > +    /// </summary>
> +    /// <param name="useSingleCollectionForVectorSearch"></param>
> +    /// <returns></returns>
> +    public MongoDbAtlasConfig WithSingleCollectionForVectorSearch(bool useSingleCollectionForVectorSearch)
> +    {
> +        this.UseSingleCollectionForVectorSearch = useSingleCollectionForVectorSearch;
> +        return this;
> +    }
> +
> +    public MongoDbAtlasConfig WithAfterIndexCallback(Func<Task> afterIndexCallback)
> +    {
> +        this.AfterIndexCallbackAsync = afterIndexCallback;
> +        return this;
> +    }
> +
> +    internal IMongoDatabase GetDatabase()
>
> the memory db object is a singleton, that should be sufficient no?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/kernel-memory/pull/311#discussion_r1518925331>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACXREOQUWR3XQTWTQXDSR3YXSXWHAVCNFSM6AAAAABDGBKWHCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMRWGYYTQMZQGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dluc commented on service/tests/Core.FunctionalTests/Core.FunctionalTests.csproj:
```diff
@@ -14,10 +14,8 @@
     </PropertyGroup>
 
     <ItemGroup>
```
No need to add project references here, I removed also old leftovers

alkampfergit replied:
Ok you are right.

alkampfergit commented on extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasDatabaseHelper.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    private static MongoClient? s_client = null;
+
+    internal static IMongoDatabase GetDatabase(string connectionString, string databaseName)
```
I changed the code, actually I've ported a code from other projects of mine, client is not static anymore but it is stored inside a static concurrent dictionary. This will ensure that we will create a single IMongoClient for each different connection string. At the same time we allow to use 2 different connections, one for storage and other for memory. 

Also used the same class from the SearchHelper class.
